### PR TITLE
JSON decoder / encoder

### PR DESCRIPTION
**Pull request by parfeon**
feat: JSON decoder / encoder

Decoder to map server response directly to the data models.

feat: request / response-based interface

Configurable request objects require less convenience methods to interact with PubNub REST API.

feat: network layer rewritten as module

### CODE REVIEW

File: Example/PubNub/PNAppDelegate.m [hunk 1]
```diff
@@ -61,8 +61,8 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
     self.channel1 = @"bot";
     self.channel2 = @"myCh";
     self.channelGroup1 = @"myChannelGroup";
-    self.pubKey = @"demo";
-    self.subKey = @"demo";
+    self.pubKey = @"pub-c-92e62c76-408a-4ac4-aefc-a1d20a83b2a6";
```
Comment (jguz-pubnub) [inline Example/PubNub/PNAppDelegate.m:64]: Is it intentional?

File: PubNub/Core/PubNub+APNS.m [hunk 2]
```diff
@@ -398,61 +209,69 @@ - (void)removeAllPushNotificationsFromDeviceWithPushToken:(id)pushToken
...
+#pragma mark - Push notifications state audit
+
+- (void)fetchPushNotificationWithRequest:(PNPushNotificationFetchRequest *)userRequest
+                              completion:(PNPushNotificationsStateAuditCompletionBlock)handlerBlock {
+    PNOperationDataParser *responseParser = [self parserWithResult:[PNAPNSEnabledChannelsResult class]
+                                                            status:[PNErrorStatus class]];
+    PNPushNotificationsStateAuditCompletionBlock block = [handlerBlock copy];
+    PNParsedRequestCompletionBlock handler;
+
+#ifndef PUBNUB_DISABLE_LOGGER
```
Comment (jguz-pubnub) [inline PubNub/Core/PubNub+APNS.m:238]: We could consider checking against the `PUBNUB_DISABLE_LOGGER` flag as a part of `PNLogAPICall`, so we would have such a check in one place only. Another idea could be overriding `debugDescription` for objects like `userRequest`. These two should shorten the overall syntax for logging. Let's leave it for now and consider pros and cons in the future

File: PubNub/Core/PubNub+Objects.m [hunk 3]
```diff
@@ -514,284 +438,487 @@ - (void)sendFetchChannelMembersRequestUsingBuilderParameters:(NSDictionary *)par
...
-            status.retryBlock = ^{
-                [weakSelf setUUIDMetadataWithRequest:request completion:block];
+- (void)setUUIDMetadataWithRequest:(PNSetUUIDMetadataRequest *)userRequest
+                        completion:(PNSetUUIDMetadataCompletionBlock)handleBlock {
+    userRequest.identifier = userRequest.identifier.length ? userRequest.identifier : self.configuration.userID;
+    PNOperationDataParser *responseParser = [self parserWithStatus:[PNSetUUIDMetadataStatus class]];
+    PNSetUUIDMetadataCompletionBlock block = [handleBlock copy];
+    PNParsedRequestCompletionBlock handler; 
+
+#ifndef PUBNUB_DISABLE_LOGGER
```
Comment (jguz-pubnub) [inline PubNub/Core/PubNub+Objects.m:448]: Looks like it's not used

File: PubNub/Core/PubNub+Objects.m [hunk 4]
```diff
@@ -514,284 +438,487 @@ - (void)sendFetchChannelMembersRequestUsingBuilderParameters:(NSDictionary *)par
...
+}
+
+- (void)removeUUIDMetadataWithRequest:(PNRemoveUUIDMetadataRequest *)userRequest
+                           completion:(PNRemoveUUIDMetadataCompletionBlock)handleBlock {
+    userRequest.identifier = userRequest.identifier.length ? userRequest.identifier : self.configuration.userID;
+    PNOperationDataParser *responseParser = [self parserWithStatus:[PNAcknowledgmentStatus class]];
+    PNRemoveUUIDMetadataCompletionBlock block = [handleBlock copy];
+    PNParsedRequestCompletionBlock handler; 
+
+#ifndef PUBNUB_DISABLE_LOGGER
```
Comment (jguz-pubnub) [inline PubNub/Core/PubNub+Objects.m:478]: The same, looks like it's not used

File: PubNub/Core/PubNub+Objects.m [hunk 5]
```diff
@@ -514,284 +438,487 @@ - (void)sendFetchChannelMembersRequestUsingBuilderParameters:(NSDictionary *)par
...
+
+- (void)uuidMetadataWithRequest:(PNFetchUUIDMetadataRequest *)userRequest
+                     completion:(PNFetchUUIDMetadataCompletionBlock)handleBlock {
+    userRequest.identifier = userRequest.identifier.length ? userRequest.identifier : self.configuration.userID;
+    PNOperationDataParser *responseParser = [self parserWithResult:[PNFetchUUIDMetadataResult class]
+                                                            status:[PNErrorStatus class]];
+    PNFetchUUIDMetadataCompletionBlock block = [handleBlock copy];
+    PNParsedRequestCompletionBlock handler; 
+
+#ifndef PUBNUB_DISABLE_LOGGER
```
Comment (jguz-pubnub) [inline PubNub/Core/PubNub+Objects.m:509]: The same, looks like it's not used

File: PubNub/Core/PubNub+Objects.m [hunk 6]
```diff
@@ -514,284 +438,487 @@ - (void)sendFetchChannelMembersRequestUsingBuilderParameters:(NSDictionary *)par
...
+}
+
+- (void)allUUIDMetadataWithRequest:(PNFetchAllUUIDMetadataRequest *)userRequest
+                        completion:(PNFetchAllUUIDMetadataCompletionBlock)handleBlock {
+    PNOperationDataParser *responseParser = [self parserWithResult:[PNFetchAllUUIDMetadataResult class]
+                                                            status:[PNErrorStatus class]];
+    PNFetchAllUUIDMetadataCompletionBlock block = [handleBlock copy];
+    PNParsedRequestCompletionBlock handler; 
+
+#ifndef PUBNUB_DISABLE_LOGGER
```
Comment (jguz-pubnub) [inline PubNub/Core/PubNub+Objects.m:539]: The same here

File: PubNub/Core/PubNub+Objects.m [hunk 7]
```diff
@@ -514,284 +438,487 @@ - (void)sendFetchChannelMembersRequestUsingBuilderParameters:(NSDictionary *)par
...
-        if (block && status.isError) {
-            status.retryBlock = ^{
-                [weakSelf removeChannelMetadataWithRequest:request completion:block];
+- (void)removeChannelMetadataWithRequest:(PNRemoveChannelMetadataRequest *)userRequest
+                              completion:(PNRemoveChannelMetadataCompletionBlock)handleBlock {
+    PNOperationDataParser *responseParser = [self parserWithStatus:[PNAcknowledgmentStatus class]];
+    PNRemoveChannelMetadataCompletionBlock block = [handleBlock copy];
+    PNParsedRequestCompletionBlock handler; 
+
+#ifndef PUBNUB_DISABLE_LOGGER
```
Comment (jguz-pubnub) [inline PubNub/Core/PubNub+Objects.m:599]: Here too, perhaps the best way would be to find all these empty occurrences and delete them

File: Example/PubNub/PNAppDelegate.m (hunk 1 above)
Comment (parfeon) [reply Example/PubNub/PNAppDelegate.m]: Nope. Fixed.

File: PubNub/Core/PubNub+Objects.m (hunk 3 above)
Comment (parfeon) [reply PubNub/Core/PubNub+Objects.m:448]: I can drop it. Just added as a placeholder (it still will be stripped away during compilation).

File: PubNub/Core/PubNub+Objects.m (hunk 7 above)
Comment (parfeon) [reply PubNub/Core/PubNub+Objects.m:599]: All Object API files has this as a placeholder. Need to sneak pick into another language which prints something and repeat the message here :)

**Review by jguz-pubnub — APPROVED** [review]

**Comment by parfeon** [comment]
@pubnub-release-bot release

**Pull request merged**